tools/Dashboard/DevHome.Dashboard/Services/WidgetIconService.cs

```diff
@@ -18,103 +17,61 @@ public class WidgetIconService : IWidgetIconService
-                var itemLightImage = await WidgetIconToBitmapImageAsync(widgetDef.GetThemeResource(WidgetTheme.Light).Icon);
-                var itemDarkImage = await WidgetIconToBitmapImageAsync(widgetDef.GetThemeResource(WidgetTheme.Dark).Icon);
-
-                // There is a widget bug where Definition update events are being raised as added events.
-                // If we already have an icon for this key, just remove and add again in case the icons changed.
-                _widgetLightIconCache.Remove(widgetDefId);
-                _widgetDarkIconCache.Remove(widgetDefId);
-
-                _widgetLightIconCache.Add(widgetDefId, itemLightImage);
-                _widgetDarkIconCache.Add(widgetDefId, itemDarkImage);
-            }
-            catch (Exception ex)
-            {
-                Log.Logger()?.ReportError("WidgetIconCache", $"Exception in AddIconsToCacheAsync:", ex);
-                _widgetLightIconCache.Add(widgetDefId, null);
-                _widgetDarkIconCache.Add(widgetDefId, null);
-            }
-        }
-    }
-
     public void RemoveIconsFromCache(string definitionId)
     {
         _widgetLightIconCache.Remove(definitionId);
         _widgetDarkIconCache.Remove(definitionId);
     }
 
-    public async Task<BitmapImage> GetWidgetIconForThemeAsync(WidgetDefinition widgetDefinition, ElementTheme theme)
+    public async Task<BitmapImage> GetIconFromCache(WidgetDefinition widgetDefinition, ElementTheme theme)
     {
-        // Return the WidgetDefinition Id via TaskCompletionSource. Using WCT's EnqueueAsync does not suffice here, since if
-        // we're already on the thread of the DispatcherQueue then it just directly calls the function, with no async involved.
-        var completionSource = new TaskCompletionSource<string>();
-        _windowEx.DispatcherQueue.TryEnqueue(() =>
+        var widgetDefinitionId = widgetDefinition.Id;
+        BitmapImage bitmapImage;
+
+        // First, check the cache to see if the icon is already there.
+        if (theme == ElementTheme.Dark)
         {
-            completionSource.TrySetResult(widgetDefinition.Id);
-        });
+            _widgetDarkIconCache.TryGetValue(widgetDefinitionId, out bitmapImage);
+        }
+        else
+        {
+            _widgetLightIconCache.TryGetValue(widgetDefinitionId, out bitmapImage);
+        }
 
-        var widgetDefinitionId = await completionSource.Task;
+        if (bitmapImage != null)
+        {
+            return bitmapImage;
+        }
 
-        BitmapImage image;
-        if (theme == ElementTheme.Light)
+        // If the icon wasn't already in the cache, get it from the widget definition and add it to the cache before returning.
+        if (theme == ElementTheme.Dark)
         {
-            _widgetLightIconCache.TryGetValue(widgetDefinitionId, out image);
+            bitmapImage = await WidgetIconToBitmapImageAsync(widgetDefinition.GetThemeResource(WidgetTheme.Dark).Icon);

```

Comment:
Can multiple threads access this method?  I'm wondering if a race-condition can happen and DevHome crashes because two threads are writing to the cache at the same time.

Reply:
Switched to ConcurrentDictionary for both icons and screenshots. Probably overkill for screenshots, but safer.